File: zgadywanko.hs
{-# LANGUAGE BlockArguments #-}

import Data.Char (isSpace, toLower)
import Data.List (delete, dropWhileEnd, partition)
import System.Environment (getArgs)
import System.IO (hFlush, stdout)
import System.Random.Stateful (uniformRM, globalStdGen)
import Text.Read (readMaybe)

replicateUniqueM :: (Monad m, Eq a) => Int -> m a -> m [a]
replicateUniqueM cnt action
  | cnt <= 0 = return []
  | otherwise = replicateUniqueM (cnt - 1) action >>= extend
  where
    extend rest = do
      num <- action
      if num `elem` rest then
        extend rest
      else
        return $ num : rest

ask :: String -> IO String
ask question = do
  putStr question
  putStr " "
  hFlush stdout
  getLine

getYesNo :: String -> IO Bool
getYesNo prompt = do
  ans <- ask $ prompt ++ " (y/n)"
  case map toLower $ dropWhileEnd isSpace $ dropWhile isSpace ans of
    "y" -> return True
    "n" -> return False
    _ -> getYesNo prompt

loop :: [Integer] -> IO ()
loop nums = do
  maybeGuess <- readMaybe <$> ask "Your guess:"
  case maybeGuess of
    Nothing -> loop nums
    Just guess ->
      let nums' = delete guess nums in
      if null nums' then
        putStrLn "You've found the last number, congratulations!"

      else do
        let didFindAny = guess `elem` nums
        putStr if didFindAny then "That's one of the numbers, good job!\n" else ""

        let (lessc, greaterc) = (\(a, b) -> (length a, length b)) $ partition (< guess) nums'
        putStr if didFindAny then "Still, too " else "Too "
        putStr case lessc of
          0 -> ""
          1 -> "high"
          x -> "high for " ++ show x
        putStr if lessc > 0 && greaterc > 0 then " and too " else ""
        putStr case greaterc of
          0 -> ""
          1 -> "low"
          x -> "low for " ++ show x
        putStrLn "!"

        loop nums'

game :: Integer -> Int -> IO ()
game maxNum numc = do
  putStrLn $ "Guess the " ++ (if numc == 1 then "number" else show numc ++ " numbers") ++ " from 1 to " ++ show maxNum ++ "."
  replicateUniqueM numc (uniformRM (1, maxNum) globalStdGen) >>= loop

  shouldContinue <- getYesNo "Do you want to play again?"
  if shouldContinue then
    game maxNum numc
  else
    return ()

main :: IO ()
main = do
  putStrLn "Welcome to The Miraculous Game of Zgadywanko!"

  args <- getArgs
  let
    (maxNum, numc) = case args of
      [] -> (100, 1)
      [a] -> (read a, 1)
      [a, b] -> (read a, read b)
      _ -> error "Too many arguments"
  if numc < 1 then
    error "The number of numbers must not be less than 1"
  else if maxNum < toInteger numc then
    error $ "The upper bound must not be less than " ++ show numc

  else do
    game maxNum numc
    putStrLn "Thanks for playing, bye!"

File: zgadywanko.clj
(require '[clojure.string :as str])

(defn get-yes-no [prompt]
  (print prompt "(y/n) ")
  (flush)
  (case (-> (read-line) str/trim str/lower-case)
    "y" true
    "n" false
    (recur prompt)))

(defn rand-big-int [a b]
  (let [result (BigInteger. (.bitLength (- b a)) (java.util.concurrent.ThreadLocalRandom/current))]
    (if (<= result (- b a)) (+ result a) (recur a b))))

(defn get-num [prompt]
  (print prompt "")
  (flush)
  (try (-> (read-line) str/trim bigint)
    (catch NumberFormatException _ (get-num prompt))))

(defn loop [nums]
  (def guess (get-num "Your guess:"))

  (def nums' (disj nums guess))
  (if (empty? nums') (println "You've found the last number, congratulations!") (do
    (def did-find-any (contains? nums guess))
    (when did-find-any (println "That's one of the numbers, good job!"))

    (def lessc (count (filter #(< % guess) nums')))
    (def greaterc (- (count nums') lessc))
    (print (if did-find-any "Still, too " "Too "))
    (when (> lessc 0) (print (if (= lessc 1) "high" (str "high for " lessc))))
    (when (and (> lessc 0) (> greaterc 0)) (print " and too "))
    (when (> greaterc 0) (print (if (= greaterc 1) "low" (str "low for " greaterc))))
    (println "!")

    (loop nums'))))

(defn game [max-num numc]
  (printf "Guess the %s from 1 to %d.\n" (if (= numc 1) "number" (str numc " numbers")) (biginteger max-num))
  (loop (->> #(rand-big-int 1N max-num) repeatedly distinct (take numc) set))
  (when (get-yes-no "Do you want to play again?") (recur max-num numc)))

(println "Welcome to The Miraculous Game of Zgadywanko!")

(when (> (count *command-line-args*) 2) (throw (Exception. "Too many arguments")))
(def max-num (nth (map bigint *command-line-args*) 0 100))
(def numc (nth (map #(Integer/parseInt %) *command-line-args*) 1 1))
(when (< numc 1) (throw (Exception. "The number of numbers must not be less than 1")))
(when (< max-num numc) (throw (Exception. (str "The upper bound must not be less than " numc))))

(game max-num numc)
(println "Thanks for playing, bye!")
